// brightspace-auth-keys-dynamodb-store/src/public-key-store.js
import { AbstractPublicKeyStore } from 'brightspace-auth-keys';

import clock from './clock.js';

class DynamoPublicKeyStore extends AbstractPublicKeyStore {
	constructor(client, tableName) {
		super();
		this._client = client;
		this._tableName = tableName;
	}

	_storePublicKey(publicKey, expiresAt) {
		return this._client.putItem(
			{
				TableName: this._tableName,
				Item: {
					'PublicKey': {
						'S': publicKey
					},
					'ExpiresAt': {
						'N': '' + expiresAt
					}
				}
			}
		).promise();
	}

	_lookupPublicKeys() {
		const parameters = {
			TableName: this._tableName,
			FilterExpression: '#e > :current_time',
			ExpressionAttributeNames: {
				'#e': 'ExpiresAt'
			},
			ExpressionAttributeValues: {
				':current_time': {
					'N': '' + clock()
				}
			},
			ConsistentRead: true
		};

		return this._scanUntilComplete(null, parameters, []);
	}

	async _scanUntilComplete(response, parameters, items) {
		if (response) {
			if (response.Items) {
				items = items.concat(response.Items.map(row => row.PublicKey.S));
			}

			if (response.LastEvaluatedKey) {
				parameters.ExclusiveStartKey = response.LastEvaluatedKey;
			} else {
				return items;
			}
		}

		const newResponse = await this._client.scan(parameters).promise();
		return this._scanUntilComplete(newResponse, parameters, items);
	}
}

export default DynamoPublicKeyStore;
export { default as createTable } from './public-key-table-setup.js';
