// brightspace-auth-provisioning/spec/provisioner.spec.js
'use strict';

const { generateKeyPairSync } = require('crypto');
const chai = require('chai');
const chaiAsPromised = require('chai-as-promised');
const sinon = require('sinon');
const { MockAgent, setGlobalDispatcher } = require('undici');

const expect = chai.expect;
chai.use(chaiAsPromised);

const AuthTokenProvisioner = require('../');
const AbstractProvisioningCache = require('../src/abstract-provisioning-cache');

const ISSUER = 'https://test.example.com/baz';
const REMOTE_ISSUER = 'http://auth.test.example.com';
const TOKEN_ENDPOINT_PATH = '/connect/token';
const MOCK_ACCESS_TOKEN = 'mock-access-token-value';

function makeKeyLookup(opts) {
const { privateKey } = generateKeyPairSync('ec', { namedCurve: 'P-256' });
return async () => Object.assign({ kid: 'test-key-1', alg: 'ES256', key: privateKey }, opts);
}

function interceptToken(cb) {
const mockAgent = new MockAgent({ connections: 1 });
const mockClient = mockAgent.get(REMOTE_ISSUER);
setGlobalDispatcher(mockClient);
cb(mockClient.intercept({
path: TOKEN_ENDPOINT_PATH,
method: 'POST',
}));
}

describe('AuthTokenProvisioner', function() {
describe('constructor', function() {
it('should throw when opts.issuer is missing', function() {
expect(() => new AuthTokenProvisioner({})).to.throw(/"opts.issuer"/);
});

it('should throw when opts.issuer is not a string', function() {
expect(() => new AuthTokenProvisioner({ issuer: 123 })).to.throw(/"opts.issuer"/);
});

it('should throw when opts.issuer is an empty string', function() {
expect(() => new AuthTokenProvisioner({ issuer: '' })).to.throw(/"opts.issuer"/);
});

it('should throw when opts.keyLookup is not a function', function() {
expect(() => new AuthTokenProvisioner({ issuer: ISSUER })).to.throw(/"opts.keyLookup"/);
});

it('should throw when opts.cache is not an AbstractProvisioningCache', function() {
expect(() => new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: async () => {},
cache: {},
})).to.throw(/"opts.cache"/);
});

it('should construct successfully with valid opts', function() {
const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: async () => {},
});
expect(provisioner).to.be.instanceof(AuthTokenProvisioner);
});

it('should use default remote issuer when not provided', function() {
const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: async () => {},
});
expect(provisioner._tokenEndpoint).to.equal('https://auth.brightspace.com/core/connect/token');
});

it('should use provided remoteIssuer', function() {
const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: async () => {},
remoteIssuer: REMOTE_ISSUER,
});
expect(provisioner._tokenEndpoint).to.equal(REMOTE_ISSUER + TOKEN_ENDPOINT_PATH);
});

it('should strip trailing slashes from remoteIssuer', function() {
const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: async () => {},
remoteIssuer: REMOTE_ISSUER + '///',
});
expect(provisioner._tokenEndpoint).to.equal(REMOTE_ISSUER + TOKEN_ENDPOINT_PATH);
});
});

describe('provisionToken', function() {
it('should throw when opts.scopes is not an array', function() {
const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: makeKeyLookup(),
remoteIssuer: REMOTE_ISSUER,
});
return expect(provisioner.provisionToken({})).to.be.rejectedWith('"opts.scopes" must be an array');
});

it('should throw when opts.user is provided without opts.tenant', function() {
const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: makeKeyLookup(),
remoteIssuer: REMOTE_ISSUER,
});
return expect(provisioner.provisionToken({ scopes: [], user: 'user1' }))
.to.be.rejectedWith('"opts.user" depends on "opts.tenant"');
});

it('should throw when opts.impersonator is provided without opts.user', function() {
const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: makeKeyLookup(),
remoteIssuer: REMOTE_ISSUER,
});
return expect(provisioner.provisionToken({
scopes: [],
tenant: 'tenant1',
impersonator: 'imp1',
})).to.be.rejectedWith('"opts.impersonator" depends on "opts.user"');
});

it('should throw when keyLookup returns an invalid key', function() {
const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: async () => null,
remoteIssuer: REMOTE_ISSUER,
});
return expect(provisioner.provisionToken({ scopes: [] }))
.to.be.rejectedWith('received invalid signing key from "keyLookup"');
});

it('should throw when keyLookup returns a key with an unsupported alg', function() {
const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: makeKeyLookup({ alg: 'HS256' }),
remoteIssuer: REMOTE_ISSUER,
});
return expect(provisioner.provisionToken({ scopes: [] }))
.to.be.rejectedWith('received invalid signing key from "keyLookup"');
});

it('should return access token on success', async function() {
interceptToken(x => x.reply(200, { access_token: MOCK_ACCESS_TOKEN }));

const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: makeKeyLookup(),
remoteIssuer: REMOTE_ISSUER,
});

const token = await provisioner.provisionToken({ scopes: ['scope:read'] });
expect(token).to.equal(MOCK_ACCESS_TOKEN);
});

it('should build a signed JWT assertion and send it to the token endpoint', async function() {
	interceptToken(x => x.reply(200, { access_token: MOCK_ACCESS_TOKEN }));

	const { privateKey } = generateKeyPairSync('ec', { namedCurve: 'P-256' });
	const provisioner = new AuthTokenProvisioner({
		issuer: ISSUER,
		keyLookup: async () => ({ kid: 'test-key-1', alg: 'ES256', key: privateKey }),
		remoteIssuer: REMOTE_ISSUER,
	});

	const makeAssertionSpy = sinon.spy(provisioner, '_makeAssertion');
	const token = await provisioner.provisionToken({ scopes: ['scope:read'] });

	expect(token).to.equal(MOCK_ACCESS_TOKEN);
	expect(makeAssertionSpy.calledOnce).to.be.true;
	const [assertion, scope] = makeAssertionSpy.firstCall.args;
	// assertion is a signed JWT (three base64url segments)
	expect(assertion).to.match(/^[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+$/);
	expect(scope).to.equal('scope:read');
});

it('should throw when token endpoint returns a non-ok response', function() {
interceptToken(x => x.reply(401, { error: 'unauthorized' }));

const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: makeKeyLookup(),
remoteIssuer: REMOTE_ISSUER,
});

return expect(provisioner.provisionToken({ scopes: [] }))
.to.be.rejectedWith('Request failed with status 401');
});

it('should attempt cache lookup before making a network request', async function() {
const getCalled = [];

class MissCache extends AbstractProvisioningCache {
async _get(key) {
getCalled.push(key);
throw new Error('cache miss');
}
}

interceptToken(x => x.reply(200, { access_token: MOCK_ACCESS_TOKEN }));

const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: makeKeyLookup(),
remoteIssuer: REMOTE_ISSUER,
cache: new MissCache(),
});

await provisioner.provisionToken({ scopes: ['scope:read'] });
expect(getCalled).to.have.length(1);
});

it('should attempt to store the token in cache after fetching', async function() {
const cache = new AbstractProvisioningCache();
const setSpy = sinon.spy(cache, 'set');

interceptToken(x => x.reply(200, { access_token: MOCK_ACCESS_TOKEN }));

const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: makeKeyLookup(),
remoteIssuer: REMOTE_ISSUER,
cache,
});

await provisioner.provisionToken({ scopes: ['scope:read'] });
expect(setSpy.calledOnce).to.be.true;
expect(setSpy.firstCall.args[2]).to.equal(MOCK_ACCESS_TOKEN);
});

it('should skip network request when cache returns a valid token', async function() {
const { privateKey } = generateKeyPairSync('ec', { namedCurve: 'P-256' });

// Build a real JWT to store in the cache.
// AbstractProvisioningCache.get decodes the JWT to check expiry,
// so a real signed JWT with a future exp is required.
const { SignJWT } = await import('jose/jwt/sign');
const nowSecs = Math.round(Date.now() / 1000);
const cachedToken = await new SignJWT({ sub: 'user1' })
.setProtectedHeader({ alg: 'ES256', kid: 'test-key-1' })
.setExpirationTime(nowSecs + 3600)
.sign(privateKey);

class HitCache extends AbstractProvisioningCache {
async _get() { return cachedToken; }
async _set() {}
}

const fetchSpy = sinon.stub(globalThis, 'fetch');
try {
const provisioner = new AuthTokenProvisioner({
issuer: ISSUER,
keyLookup: makeKeyLookup(),
remoteIssuer: REMOTE_ISSUER,
cache: new HitCache(),
});

const result = await provisioner.provisionToken({ scopes: ['scope:read'] });
expect(result).to.equal(cachedToken);
expect(fetchSpy.called).to.be.false;
} finally {
fetchSpy.restore();
}
});
});

describe('AbstractProvisioningCache export', function() {
it('should export AbstractProvisioningCache', function() {
expect(AuthTokenProvisioner.AbstractProvisioningCache).to.equal(AbstractProvisioningCache);
});
});
});
